components/text_recognition/browser/text_recognition_win.cc

```diff
@@ -0,0 +1,230 @@
+  hr = GetActivationFactory<ILanguageFactory,
+                            RuntimeClass_Windows_Globalization_Language>(
+      &language_factory);
+  if (FAILED(hr)) {
+    VLOG(2) << "ILanguage factory failed: "
+            << logging::SystemErrorCodeToString(hr);
+    RunOnUIThread(std::move(callback_run_on_ui_thread));
+    return;
+  }
+
+  ComPtr<IOcrEngine> ocr_engine;
+  if (language_code.empty()) {
+    hr = engine_factory->TryCreateFromUserProfileLanguages(&ocr_engine);
+    if (FAILED(hr)) {
+      VLOG(2) << "Create engine failed from language: "
+              << logging::SystemErrorCodeToString(hr);
+      RunOnUIThread(std::move(callback_run_on_ui_thread));
+      return;
+    }
+  } else {
+    ScopedHString language_hstring = ScopedHString::Create(language_code);
+    if (!language_hstring.is_valid()) {
+      RunOnUIThread(std::move(callback_run_on_ui_thread));
+      return;
+    }
+
+    ComPtr<ILanguage> language;
+    hr = language_factory->CreateLanguage(language_hstring.get(), &language);
+    if (FAILED(hr)) {
+      VLOG(2) << "Create language failed: "
+              << logging::SystemErrorCodeToString(hr);
+      RunOnUIThread(std::move(callback_run_on_ui_thread));
+      return;
+    }
+
+    boolean is_supported = false;
+    hr = engine_factory->IsLanguageSupported(language.Get(), &is_supported);
+    if (FAILED(hr) || !is_supported) {
+      RunOnUIThread(std::move(callback_run_on_ui_thread));
+      return;
+    }
+
+    hr = engine_factory->TryCreateFromLanguage(language.Get(), &ocr_engine);
+    if (FAILED(hr)) {
+      VLOG(2) << "Create engine failed from language: "
+              << logging::SystemErrorCodeToString(hr);
+      RunOnUIThread(std::move(callback_run_on_ui_thread));
+      return;
+    }
+  }
+
+  ComPtr<ISoftwareBitmapStatics> bitmap_factory;
+  hr = GetActivationFactory<
+      ISoftwareBitmapStatics,
+      RuntimeClass_Windows_Graphics_Imaging_SoftwareBitmap>(&bitmap_factory);
+  if (FAILED(hr)) {
+    VLOG(2) << "ISoftwareBitmapStatics factory failed: "
+            << logging::SystemErrorCodeToString(hr);
+    RunOnUIThread(std::move(callback_run_on_ui_thread));
+    return;
+  }
+
+  auto recognizer = std::make_unique<TextRecognizerWin>(
+      std::move(ocr_engine), std::move(bitmap_factory));
+  auto* recognizer_ptr = recognizer.get();
+  recognizer_ptr->Detect(
+      image, base::BindOnce(
+                 [](std::unique_ptr<TextRecognizerWin> recognizer,
+                    base::OnceCallback<void(const std::vector<std::string>&)>
+                        callback_run_on_ui_thread,
+                    std::vector<shape_detection::mojom::TextDetectionResultPtr>
+                        result) {
+                   std::vector<std::string> detected_string;
+                   for (const auto& re : result) {
+                     detected_string.push_back(re->raw_value);
+                   }
+                   RunOnUIThread(std::move(callback_run_on_ui_thread),
```

Comment:
1. let's remove `RunOnUIThread` and pass the "reply" task runner as a parameter into `TextRecognizerWin`. You don't need to depend on `content` for that, this should simplify things.
2. you can split `GetTextFromImage` into `init` and `detect` parts. The `init` part can just return `true/false` so you don't have to do `RunOnUIThread(std::move(callback_run_on_ui_thread));` in each `FAILED(hr)` case.

Reply:
Removed `content` dependency and make `GetTextFromImage` gets `reply_runner` as args.
and make `GetTextFromImage` return `false` when text recognization is not available.
Seems much cleaner :)